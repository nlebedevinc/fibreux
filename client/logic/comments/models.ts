/**
 * This is a definition of domain model for `comments` context.
 *
 * Here we only write classes and types that are shared with the server.
 * In case you need a type that only work on the client,
 * then use a regular typescript type and place it in `/types.ts` file.
 */

import * as ts from 'io-ts'

/**
 * Runtime type, that can be used for schema validation.
 *
 * We call them "models", because they are exchanged between client and server.
 *
 * @see https://github.com/gcanti/io-ts
 * @see https://github.com/aeirola/io-ts-promise
 */
export const RawComment = ts.type({
  'id': ts.number,
  'body': ts.string,
  'email': ts.string,
})

// Static TypeScript type, that can be used as a regular `type`:
export type RawCommentType = ts.TypeOf<typeof RawComment>
