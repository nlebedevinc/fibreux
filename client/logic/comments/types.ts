/**
 * These are inner types, they are only used inside the client.
 *
 * If your exchange some information with a server - make it a model.
 * And place it in `/models.ts` file.
 */

import { RawCommentType } from '~/logic/comments/models'

// We add rating only on the client (for demo purposes)
export type CommentType = RawCommentType & {
  rating: number
}

export interface CommentPayloadType {
  commentId: number
  delta: number
}
