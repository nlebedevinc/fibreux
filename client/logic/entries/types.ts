// This is a definition of domain model for `entries` context.

import * as ts from 'io-ts'

// Runtime type, that can be used for schema validation:
export const Entry = ts.type({
  'id': ts.string,
  'description': ts.union([ts.string, ts.null]),
  'person': ts.string,
  'when': ts.union([ts.string, ts.null]),
  'time': ts.union([ts.string, ts.null]),
  'ticket': ts.union([ts.string, ts.null]),
  'project': ts.string,
})

export enum Filters {
  Day = 1,
  Week,
  Month,
}

export const Period = ts.type({
  'from': ts.string,
  'to': ts.string,
})

export type PeriodType = ts.TypeOf<typeof Period>

// Static TypeScript type, that can be used as a regular `type`:
export type EntryType = ts.TypeOf<typeof Entry>

export interface StateType {
  entries: EntryType[]
  activeFilter: number
  activeDate: Date
  selectedEntry: EntryType | null
}
