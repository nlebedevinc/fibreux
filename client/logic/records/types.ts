// This is a definition of domain model for `comments` context.

import * as ts from 'io-ts'

// Runtime type, that can be used for schema validation:
export const RawComment = ts.type({
  'id': ts.number,
  'body': ts.string,
  'email': ts.string,
})

// Static TypeScript type, that can be used as a regular `type`:
export type RawCommentType = ts.TypeOf<typeof RawComment>

// We add rating only on the client (for demo purposes)
export type CommentType = RawCommentType & {
  rating: number;
}

export interface CommentPayloadType {
  commentId: number;
  delta: number;
}

export interface StateType {
  comments: CommentType[]
}
