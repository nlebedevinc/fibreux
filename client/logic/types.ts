import { CommentType } from '~/logic/comments/types'

/**
 * We use this state type to enforce the same structure with `vuex-simple`.
 *
 * There are places where we use `Store<StateType>` and it cannot be changed.
 * This is a temporary solution.
 *
 * @see https://github.com/sascha245/vuex-simple/issues/30
 */
export interface RootStateType {
  // Module:
  comments: {
    // Module's state:
    comments: CommentType[]
  }
}
