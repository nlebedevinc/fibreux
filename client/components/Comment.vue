<template>
  <div :class="computedClasses">
    <p :class="$style.author">
      {{ comment.email }}
    </p>

    <p :class="$style.body">
      {{ comment.body }}
    </p>

    <div :class="$style.rating">
      <button
        :class="$style['rating-up']"
        @click="changeRating(comment.id, 1)"
        @keypress.enter.prevent="changeRating(comment.id, 1)"
      >
        +
      </button>

      <span :class="$style.number">{{ comment.rating }}</span>

      <button
        :class="$style['rating-down']"
        @click="changeRating(comment.id, -1)"
        @keydown.enter.prevent="changeRating(comment.id, -1)"
      >
        -
      </button>
    </div>
  </div>
</template>

<script lang="ts">
import Component, { mixins } from 'nuxt-class-component'
import { Prop } from 'vue-property-decorator'

import { CommentType } from '~/logic/comments/types'
import TypedStoreMixin from '~/mixins/typed-store'

// @vue/component
@Component({})
/**
 * Comment component is used to represent a single user's comment.
 */
export default class Comment extends mixins(TypedStoreMixin) {
  @Prop({ 'required': true })
  /**
   * Passed comment from the parent component.
   */
  readonly comment!: CommentType

  /**
   * Changes comment's rating.
   * Can be used to increase or decrease comment's rating.
   *
   * @param commentId - Comment's identifier to change rating.
   * @param delta - Delta value to change rating value.
   */
  changeRating(commentId: number, delta: number): void {
    this.typedStore.comments.updateRating({ commentId, delta })
  }

  /**
   * Defines which color borders are.
   * Uses `comment`'s rating to choose color.
   *
   * @returns Pairs of class names and boolean values if they should be applied.
   */
  get computedClasses(): Readonly<Record<string, boolean>> {
    return {
      [this.$style['comment-component']]: true,
      [this.$style['comment-positive']]: this.comment.rating > 0,
      [this.$style['comment-negative']]: this.comment.rating < 0,
    }
  }
}
</script>

<style lang="scss" module>
@import '~/scss/variables';

.comment-component {
  background: $color-white;
  max-width: 15.6rem;
  border: 0.12rem solid;
  padding: 0.74rem;
  margin: 0.5rem;
  position: relative;
}

.comment-positive {
  border-color: $button-color-green;
}

.comment-negative {
  border-color: $button-color-red;
}

.author {
  padding-bottom: 0.74rem;
  font-weight: 700;
  border-bottom: 0.12rem solid;
}

.body {
  text-align: justify;
}

.rating {
  position: absolute;
  bottom: 0.3rem;
  left: 50%;
  // stylelint-disable-next-line plugin/no-unsupported-browser-features
  transform: translateX(-50%);

  .number {
    // This will be available as `this.$style.number`:
    font-weight: 700;
  }

  button {
    margin: 0 0.5rem;
    border: none;
    cursor: pointer;
    width: 2rem;
  }
}

.rating-up {
  background: $button-color-green;
}

.rating-down {
  background: $button-color-red;
}
</style>
