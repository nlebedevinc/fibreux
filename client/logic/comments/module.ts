import { Inject, Injectable } from 'vue-typedi'
import { Action, Mutation, State, Getter } from 'vuex-simple'

import { RawCommentType } from '~/logic/comments/models'
import CommentService from '~/logic/comments/services/api'
import { CommentType, CommentPayloadType } from '~/logic/comments/types'
import tokens from '~/logic/tokens'

@Injectable()
/**
 * Represents a typed Vuex module.
 *
 * @see https://vuex.vuejs.org/guide/modules.html
 * @see https://github.com/sascha245/vuex-simple
 */
export default class CommentsModule {
  // Dependencies

  @Inject(tokens.COMMENT_SERVICE)
  public service!: CommentService

  // State

  @State()
  public comments: CommentType[] = []

  // Getters

  @Getter()
  public get hasComments(): boolean {
    return Boolean(this.comments && this.comments.length > 0)
  }

  // Mutations

  @Mutation()
  public setComments(payload: RawCommentType[]): void {
    const updatedComments: CommentType[] = []

    for (const comment of payload.slice(0, 10)) {
      // We transform RawCommentType in CommentType here:
      updatedComments.push({ ...comment, 'rating': 0 })
    }

    this.comments = updatedComments
  }

  @Mutation()
  public updateRating({ commentId, delta }: CommentPayloadType): void {
    if (!this.comments) return

    const commentIndex = this.comments.findIndex((comment): boolean => {
      return comment.id === commentId
    })

    if (!this.comments || !this.comments[commentIndex]) return

    this.comments[commentIndex].rating += delta
  }

  // Actions

  @Action()
  public async fetchComments(): Promise<RawCommentType[]> {
    const commentsList = await this.service.fetchComments()
    this.setComments(commentsList)
    return commentsList
  }
}
