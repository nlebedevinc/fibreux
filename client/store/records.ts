import { actions } from '~/logic/records/actions'
import { getters } from '~/logic/records/getters'
import { mutations } from '~/logic/records/mutations'
import { state } from '~/logic/records/state'

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations,
}
