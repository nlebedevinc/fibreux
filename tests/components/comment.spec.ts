import { mount, createLocalVue } from '@vue/test-utils'
import Vue from 'vue'
import { Store } from 'vuex'

import Comment from '~/components/Comment.vue'
import { CommentType } from '~/logic/comments/types'
import TypedStore from '~/logic/store'
import { StateType } from '~/logic/records/types'

import { storeFactory, commentFactory } from '@/fixtures/vuex'

const localVue = createLocalVue()

describe('unit tests for Comment component', () => {
  let comment: CommentType
  let store: Store<StateType>

  beforeEach(() => {
    comment = commentFactory.build({ 'rating': 0 })
    store = storeFactory.build(undefined, {
      localVue,
      'state': { 'comments': { 'comments': [comment] } },
    })
  })

  test('should have valid props', () => {
    expect.hasAssertions()

    expect(Comment).toBeValidProps({ comment }, {
      localVue,
      store,
    })
  })

  test('should have two buttons', () => {
    expect.hasAssertions()

    const wrapper = mount(Comment, {
      store,
      localVue,
      'propsData': { comment },
    })

    expect(Boolean(wrapper.vm)).toBe(true)
    expect(wrapper.findAll('button')).toHaveLength(2)
  })

  test('should have correct values', () => {
    expect.hasAssertions()

    const wrapper = mount(Comment, {
      store,
      localVue,
      'propsData': { comment },
    })

    expect(wrapper.find('.body').text().trim()).toStrictEqual(comment.body)
    expect(wrapper.find('.author').text().trim()).toStrictEqual(comment.email)
    expect(wrapper.find('.number').text().trim())
      .toStrictEqual(comment.rating.toString())
  })

  test.each([
    { 'delta': 1, 'styleName': 'comment-positive' },
    { 'delta': -1, 'styleName': 'comment-negative' },
  ])('should change rating: %p', async ({ delta, styleName }) => {
    expect.hasAssertions()

    const newRating = comment.rating + delta
    const wrapper = mount(Comment, {
      store,
      localVue,
      'propsData': { comment },
    })
    const typedStore: TypedStore = (wrapper.vm as any).typedStore

    typedStore.comments.updateRating({ 'commentId': comment.id, delta })

    // This is required, because we use new version of
    // vue-test-utils, which does not have a `sync` mode to render UI:
    await Vue.nextTick()

    expect(wrapper.vm.$store.state.comments.comments[0].rating)
      .toStrictEqual(newRating)
    expect(typedStore.comments.comments[0].rating)
      .toStrictEqual(typedStore.comments.comments[0].rating)

    expect(wrapper.props().comment.rating).toStrictEqual(newRating)
    expect(wrapper.classes()).toContain(wrapper.vm.$style[styleName])
  })
})

describe('snapshot tests for Comment component', () => {
  let comment: CommentType
  let store: Store<StateType>

  beforeAll(() => {
    // We need a seed here to be consistent for snapshot testing:
    comment = commentFactory.build({}, { 'seed': 8874 })
    store = storeFactory.build(undefined, {
      localVue,
      'state': { 'comments': { 'comments': [comment] } },
    })
  })

  test('should match the snapshot', () => {
    expect.hasAssertions()

    const wrapper = mount(Comment, {
      store,
      localVue,
      'propsData': { comment },
    })

    expect(wrapper).toMatchSnapshot()
  })
})
