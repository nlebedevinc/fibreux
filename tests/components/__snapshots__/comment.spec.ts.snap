// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests for Comment component should match the snapshot 1`] = `
<div class="comment-component comment-positive">
  <p class="author">
    Mellie_Gibson59@hotmail.com
  </p>
  <p class="body">
    Sequi provident inventore dicta. Quia explicabo assumenda eum.
  </p>
  <div class="rating"><button class="rating-up">
      +
    </button> <span class="number">9</span> <button class="rating-down">
      -
    </button></div>
</div>
`;
